File: Problem450/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else if root.Val < key {
		root.Right = deleteNode(root.Right, key)
	} else {
		if root.Left == nil && root.Right == nil {
			return nil
		} else if root.Left == nil {
			return root.Right
		} else if root.Right == nil {
			return root.Left
		} else {
			left := root.Left
			for left.Right != nil {
				left = left.Right
			}
			left.Right = root.Right
			return root.Left
		}
	}
	return root
}

File: Problem450/Solution.java
import Solution.TreeNode;

public class Solution
{
    public TreeNode deleteNode(TreeNode root, int key)
    {
        if (root == null)
            return null;
        if (root.val > key)
        {
            root.left = deleteNode(root.left, key);
        }
        else if (root.val < key)
        {
            root.right = deleteNode(root.right, key);
        }
        else
        {
            if (root.left == null && root.right == null)
            {
                return null;
            }
            else if (root.right == null)
            {
                return root.left;
            }
            else if (root.left == null)
            {
                return root.right;
            }
            else
            {
                TreeNode left = root.left;
                while (left.right != null)
                {
                    left = left.right;
                }
                left.right = root.right;
                return root.left;
            }

        }
        return root;
    }
}